File: c++/039.cpp
class Solution {
    vector<vector<int>> res;
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        dfs({},candidates,target,0);
        return res;
    }
    void dfs(vector<int> r, vector<int>& c, int t, int k){  //因为调用函数时r的实参是临时的，所以不能引用&
        int sum = accumulate(r.begin(),r.end(),0);
        if(sum==t){
            res.push_back(r);
            return;
        }else if(sum>t)
                return;
        for(int i=k;i<c.size();i++){
            r.push_back(c[i]);
            dfs(r,c,t,i);
            r.pop_back();
        }
    }
};

File: go/039.go
package main

import "sort"

/**
组合总和  回溯 + 剪枝
*/
func combinationSum(candidates []int, target int) [][]int {
	if len(candidates) == 0 {
		return [][]int{}
	}
	var c []int
	var res [][]int
	sort.Ints(candidates)
	dfsSum(candidates, target, 0, c, &res)
	return res
}

func dfsSum(nums []int, target, index int, c []int, res *[][]int) {
	if target <= 0 {
		if target == 0 {
			b := make([]int, len(c))
			copy(b, c)
			*res = append(*res, b)
		}
		return
	}
	for i := index; i < len(nums); i++ {
		if target-nums[i] < 0 {
			break
		}
		c = append(c, nums[i])
		dfsSum(nums, target-nums[i], i, c, res)
		c = c[:len(c)-1]
	}
}
